### PR TITLE
Add ENABLE_BUILT_IN_PLUGINS environment variable

### PR DESCRIPTION
**Pull request by Jamalarm**
Hey Guys,

I've added some lightweight logic to the `docker-entrypoint.sh` file to allow easy configuration and usage of the plugins that come pre-packaged with the Flink distribution. For our use-case specifically we need the S3 Filesystem plugin for our AWS ECS-hosted cluster.

To use the new variable, you provide a list of all the jars you want moved from the `flink/opt` folder to the `flink/plugins` folder, delimited by semicolons. So, for example: 

```
ENABLE_BUILT_IN_PLUGINS=flink-s3-fs-presto-1.9.1.jar;flink-s3-fs-hadoop-1.9.1.jar
```

If you don't provide the variable, nothing happens.
If you provide the variable but point to a jar NOT present in `flink/opt` then the container will fail to start.

I would really appreciate a speedy review and merge of this as it is currently blocking us. I am ready to make any changes required at the drop of a hat. Just let me know.

Tom

### CODE REVIEW

**Comment by elanv** [comment]
It looks very useful. 
If it's possible to use symlinks for plugins, that also looks good.

**Comment by Jamalarm** [comment]
@pangliang are you able to take a look at this? It would be incredibly useful if we could merge it in soon, otherwise I'm going to have maintain forked docker containers to access this functionality.

**Comment by Jamalarm** [comment]
@patricklucas Apologies to keep chasing but it would be super super useful if we could merge this as soon as possible. Could you please take a look?

**Comment by patricklucas** [comment]
Hi @Jamalarm, thank you for the contribution. I would like to get input from a Flink committer or two about this, to make sure any changes we make here complement Flink itself when it comes to the plugins feature.

Additionally, there is currently a vote on the Flink Dev mailing list[1] for moving this repo into Apache which will link it more closely with the Flink release process. This may impact how fast this can be merged, but hopefully shouldn't add too much delay.

[1]https://lists.apache.org/thread.html/r13905ac2b0a03af58370d0d1dcf7c286440f8e165004ffc7cfe7c540%40%3Cdev.flink.apache.org%3E

File: docker-entrypoint.sh
Comment (AHeise) [inline docker-entrypoint.sh]: Copying the plugins directly into Flink's plugins folder is unfortunately not supported yet. You need to create a subfolder for each plugin. See also https://ci.apache.org/projects/flink/flink-docs-master/ops/plugins.html#isolation-and-plugin-structure

File: docker-entrypoint.sh
Comment (AHeise) [inline docker-entrypoint.sh]: I'm not really familiar with Docker, but please double-check that this is not modifying the base image. We need proper isolation between two different invocations of docker with different `ENABLE_BUILT_IN_PLUGINS`, especially for s3 where two plugins serve the same prefix. 
My guess is that the entrypoint is not modifying the layer and all is good.

**Review by AHeise — CHANGES_REQUESTED** [review]
Thank you very much for your contribution. It seems like a very valuable addition.
Plugin structure is not as expected yet, except for that LGTM.
Did you run any tests on that?

File: docker-entrypoint.sh
Comment (Jamalarm) [reply docker-entrypoint.sh]: Looking at the plugins currently shipped in the `opt` directory, I don't see any that need to be bundled with another jar in order to work. Therefore, are you ok with me placing each plugin in a directory with the same name as the jar? So: `flink/plugins/flink-s3-fs-presto-1.8.1/flink-s3-fs-presto-1.8.1.jar` ?

File: docker-entrypoint.sh
Comment (Jamalarm) [reply docker-entrypoint.sh]: Yep, the entrypoint **script** will be changed in the base images pushed to Docker Hub, but the entrypoint is only invoked at runtime. Someone using the container without the environment variable set will not see any changes whatsoever.

File: docker-entrypoint.sh
Comment (Jamalarm) [reply docker-entrypoint.sh]: I've pushed this change - let me know what you think

**Review by AHeise — APPROVED** [review]
LGTM

**Comment by Jamalarm** [comment]
The Travis build seemed to fail, but as far as I can see it was due to an unrelated DNS error. I'm not sure how to rerun it...

**Comment by patricklucas** [comment]
> The Travis build seemed to fail, but as far as I can see it was due to an unrelated DNS error. I'm not sure how to rerun it...

The failure was likely due to the release of Flink 1.9.2 which means the 1.9.1 artifacts are no longer available from that URL. #95 should be merged soon, then you can rebase with master and the build will pass.

File: docker-entrypoint.sh
Comment (patricklucas) [inline docker-entrypoint.sh]: Could you add a check before this line to no-op if no plugins are specified?

```bash
if [ -z "$ENABLE_BUILT_IN_PLUGINS" ]; then
    return 0
fi
```

File: docker-entrypoint.sh
Comment (patricklucas) [inline docker-entrypoint.sh]:
```suggestion
    echo "Moving ${TARGET_PLUGIN} to plugin directory"
```

(style convention of this file)

File: docker-entrypoint.sh
Comment (patricklucas) [inline docker-entrypoint.sh]:
```suggestion
  for target_plugin in $(echo "$ENABLE_BUILT_IN_PLUGINS" | grep -o -e "[^;]*"); do
```

Lowercase to distinguish from globals.

File: docker-entrypoint.sh
Comment (patricklucas) [inline docker-entrypoint.sh]:
```suggestion
    plugin_name="${TARGET_PLUGIN%.jar}"
```

Lowercase to distinguish from globals and quoting assignment.

File: docker-entrypoint.sh
Comment (patricklucas) [reply docker-entrypoint.sh]: (also in `echo` strings below)

**Review by patricklucas — CHANGES_REQUESTED** [review]
I mentioned a few style changes and one suggestion to avoid output if a user isn't using this feature.

The only breaking issue I see is that this will fail on container restart on `mkdir` since the plugin dir will already exist. That can be solved with `mkdir -p`, though you may want to consider just `continue`-ing if the destination file already exists.

Edit: actually, specifying multiple plugins doesn't work for me. It's possibly a `grep` variant discrepancy. Consider the IFS method (e.g. https://stackoverflow.com/a/918931) or using AWK.

**Comment by Jamalarm** [comment]
> Edit: actually, specifying multiple plugins doesn't work for me. It's possibly a `grep` variant discrepancy. Consider the IFS method (e.g. https://stackoverflow.com/a/918931) or using AWK.

This seems a bit odd to me. My understanding of docker was that there was no variance at all between different containers using the same image, I don't see how we could end up with different `grep` implementations...

I just ran a quick test with `ENABLE_BUILT_IN_PLUGINS=flink-s3-fs-presto-1.9.1.jar;flink-cep_2.11-1.9.1.jar;flink-metrics-datadog-1.9.1.jar` and it seemed fine. Could you send me an example where it doesn't work for you?

**Comment by patricklucas** [comment]
> This seems a bit odd to me. My understanding of docker was that there was no variance at all between different containers using the same image, I don't see how we could end up with different `grep` implementations...

That's true, but I would hope to come up with an entrypoint script that can be run outside of the context of the particular base image we're using, for testing and in general (i.e. is more standards compliant).

This is a minimal test that I ran on MacOS, where the loop only runs once instead of three times:

```bash
$ cat docker-entrypoint.sh
#!/bin/sh

export FLINK_HOME=asdf
export ENABLE_BUILT_IN_PLUGINS='foo;bar;baz'

echo "$ENABLE_BUILT_IN_PLUGINS"

copy_plugins_if_required() {
  echo "Enabling required built-in plugins"
  for TARGET_PLUGIN in $(echo "$ENABLE_BUILT_IN_PLUGINS" | grep -o -e "[^;]*"); do
    echo "Moving $TARGET_PLUGIN to plugin directory"
  done
}

copy_plugins_if_required

$ ./docker-entrypoint.sh
foo;bar;baz
Enabling required built-in plugins
Moving foo to plugin directory
```

**Comment by Jamalarm** [comment]
I'm looking at the other options for splitting, and a lot of them seem to be very distribution specific anyway. Are the `awk` or `IFS` approaches inherently more portable? I can see some comments suggesting that `awk` implementations differ over linux distributions as well, and especially with MacOS.

I appreciate the principle of designing the entrypoint to be portable, but is that not kind of the point of docker in the first place? The fact that it is dockerized means that it is 100% portable to any platform (including ones that have no notion of `grep` or `awk` at all).

**Comment by patricklucas** [comment]
AFAIK `IFS` is POSIX, so is probably the best option even though it's fairly ugly. In this case, I think it's reasonable to ask that I can verify the behavior of this otherwise-standards-compliant script on my machine without always running it in Docker.

Another note: this behavior will hopefully be added to the Flink entrypoint scripts themselves at some point (as discussed in [FLINK-14812](https://issues.apache.org/jira/browse/FLINK-14812)), which is definitely preferred, but I'm still okay with including your solution for Flink 1.9/1.10. If you would like to have a say about the longer-term solution, I encourage you to comment on that issue (and maybe look into contributing code). Once this is merged, I plan to update the docker-flink docs to clarify the behavior of these images that can change between Flink versions (such as this, or e.g. environment substitution in the config file).

**Comment by patricklucas** [comment]
Oh, one last question: would symlinks work equivalently to copying the JARs? If so, I think we might as well save the bytes in the container's writable layer with `ln -fs` instead of `cp`.

**Comment by Jamalarm** [comment]
> AFAIK `IFS` is POSIX, so is probably the best option even though it's fairly ugly. In this case, I think it's reasonable to ask that I can verify the behavior of this otherwise-standards-compliant script on my machine without always running it in Docker.

FYI - implementing the suggested IFS-based solution resulted in multiple warnings in IntelliJ, apparently `read -a` and `array references` are undefined in POSIX. I can still try and get that approach working if you like but I don't think it's going to get us closer to a standard script that can run in any shell.

> Oh, one last question: would symlinks work equivalently to copying the JARs? If so, I think we might as well save the bytes in the container's writable layer with ln -fs instead of cp.

I think this is fine, I'll try and get it working.

**Comment by patricklucas** [comment]
Since we're relying on shell tokenization anyway, what do you think about just using `tr` to replace `;` with a space? It's simple and gets the job done.

```bash
$ cat script.sh
#!/bin/sh

ENABLE_BUILT_IN_PLUGINS='foo;bar;baz asdf'

for thing in $(echo "$ENABLE_BUILT_IN_PLUGINS" | tr ';' ' '); do
  echo "here is a ${thing}"
done

$ ./script.sh
here is a foo
here is a bar
here is a baz
here is a asdf
```

AFAIK `tr` is pretty maximally compatible, and this is its most basic usage.

**Comment by Jamalarm** [comment]
`tr` looks good to me, I've made the changes and it seems to work just as well as `grep` and IntelliJ isn't complaining about POSIX issues.

**Comment by patricklucas** [comment]
Thanks, I'm testing locally and think we can merge it soon.

Unfortunately, `ln -fs` succeeds even when target file doesn't exist, unlike `cp`. My suggestion is to change:

```bash
    if mkdir -p "${FLINK_HOME}/plugins/${plugin_name}" && ln -fs "${FLINK_HOME}/opt/${target_plugin}" "${FLINK_HOME}/plugins/${plugin_name}"; then
      echo "Successfully enabled ${target_plugin}"
    else
      echo "Failed to enable ${target_plugin}. Exiting."
      exit 1
    fi
```

to:

```bash
  for target_plugin in $(echo "$ENABLE_BUILT_IN_PLUGINS" | tr ';' ' '); do
    echo "Linking ${target_plugin} to plugin directory"
    plugin_name=${target_plugin%.jar}

    mkdir -p "${FLINK_HOME}/plugins/${plugin_name}"
    if [ ! -e "${FLINK_HOME}/opt/${target_plugin}" ]; then
      echo "Plugin ${target_plugin} does not exist. Exiting."
    fi

    ln -fs "${FLINK_HOME}/opt/${target_plugin}" "${FLINK_HOME}/plugins/${plugin_name}"
    echo "Successfully enabled ${target_plugin}"
  done
```

What do you think?

**Comment by Jamalarm** [comment]
> What do you think?

Looks good to me. I tweaked it slightly so the `exit 1` call was still included, and I put the other stuff in an `else` branch which is just a personal preference. I've tested in docker with existing and non-existing files.

**Comment by patricklucas** [comment]
> > What do you think?
> 
> Looks good to me. I tweaked it slightly so the `exit 1` call was still included, and I put the other stuff in an `else` branch which is just a personal preference. I've tested in docker with existing and non-existing files.

I editing my comment when I realized it was missing but you were too quick. :)

I'll give it one more shot and then merge. Just under the wire too, I'm about to open the PR to migrate this repo to https://github.com/apache/flink-docker!

**Comment by patricklucas** [comment]
I just had the thought that `:` would be a more typical separator for JARs than `;`, but that's easy enough to add to the `tr` later if we want.

For now I'm going to merge this. Thanks a lot for the contribution!

**Review by patricklucas — APPROVED** [review]

**Comment by Jamalarm** [comment]
@patricklucas when do you think we'll see the updated entrypoint get deployed to the official images on Docker Hub? I see they were updated recently but they're still containing the older version of the script.

**Comment by patricklucas** [comment]
@Jamalarm check out the issue I linked earlier today (docker-library/official-images#7398). Once that's merged—typically within a day—it usually doesn't take long for the updated images to become available.

**Pull request merged**
